lib/cfg.c

```diff
@@ -365,7 +399,18 @@ cfg_init(GlobalConfig *cfg)
   log_template_options_init(&cfg->template_options, cfg);
   if (!cfg_init_modules(cfg))
     return FALSE;
-  return cfg_tree_start(&cfg->tree);
+
+  if (!cfg_tree_start(&cfg->tree))
+    return FALSE;
+
+  MainLoop *main_loop = main_loop_get_instance();
+  if (main_loop_get_options(main_loop) && main_loop_get_options(main_loop)->config_graph_file)
+    {
+      write_config_graph_to_file(&cfg->tree,
```

Comment:
I guess user would do it anyway, what do you think about instead making it a syslog-ng option creating a query via syslog-ng-ctl? That way you can ask for the graph later phase, where potentially additional pipe could come and go.

Reply:
That also occured to me. And it would be a nice improvement. The reason I did not do that because this time of writing there is no driver that changes the graph after init is called. Besides, adding a new ctl command would have been a little more complex than what I did.

While I agree it would be better put this into ctl, I would not do this now. The changeset is very complex as it is, and I am trying to find a sweet spot of adding an usable functionality and reasonable feature cut.

Later the ctl-time config graph dump can be added incrementally. Either by removing or leaving the start-time dump. I think we do not need to maintain backward compatibility in this feature.

Reply:
For me it feels like a more natural evolution of first providing the ctl (with cached pipe structure, just like preprocessed config is done), and later iterate over that to resolve issue why this config dump does not work during running phase.

While I still would prefer the above iteration over the current, I can and do accept yours as well.

Reply:
Ok I see. I will think about it. Let's keep this question open.

Reply:
Fixed: from now on graph is generated by ctl.